### ISSUE TITLE
Release 93u+m/1.0.5

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#619]
To-do list for the 1.0.5 release.

### High prio:

* Update `TODO` on the 1.0 branch
* Update `ANNOUNCE` on the 1.0 branch

### If feasible:

* Fix crashing bug in #435 

### Done:
* Fix or at least semi-fix #269, if possible (please test latest patch)
  √ ***Done:*** 1731f66e7bad70b35ee73e6f16426f351fd28003 (1.0), 92f65cb8086e35f3fd84fb663b24f32fafd855e3 (dev)
* Fix #616 (please test latest patch)
  √ ***Done:*** 0af49411eace399b5e4d8e2d27b28266b1c0c7a7 (1.0), d446015efce83f7cd17da830df41f8d27f10ab07 (dev)
* 9c037b0458dc44b37f7f8ca773b307e76b41e6b1 didn't actually fix #427, it just made it harder to trigger. ```fn=([foo_key]=foo_val [bar_key\]=])``` still results in ```-ksh: [bar_key]=]: not found``` instead of the expected syntax error. Fix it properly.
  √ ***Done:*** cbd7e802784280b5b98b7e7cac672551c2d698b0 (1.0), 4405a3527feae43fbbb72e823169a9cac8744fa1 (dev)
* Fix #620 
  √ ***Done:*** 1eb0b066be0c3eba8b4fe6630db1f3c1fd6dfc29 (1.0), dac884f78d90389a2bb1f9823fbf50802b5ba5d1 (dev)
* Test and merge #617 to fix #324 
  √ ***Done:*** 30a16991eff4e06eadf6e955f0921e79d192674d (1.0), 87480ec207a56d7b17de722789437c8cc7827495 (dev)
* Re 887b043f465c7c72a303d1636c1909abcd6254d2: There are more places that call nv_putsub() (outside of varsub()), and there are also two other functions that call array_grow(): array_find() and nv_atypeindex(). Check around those calls for potentially invalidated array pointers.
    √ Manual check done; as far as I can tell there are no remaining issues, though I could of course have missed some.

**Comment by McDutchie** [comment]
Ping @phidebian, @hyenias: I really want to release 1.0.5 within a week or so as it's badly overdue; it's been more than half a year since the last point release and a load of fixes have been applied since. But I'd really like to merge #569 and #617 if possible.

* #569 still has [a problem](https://github.com/ksh93/ksh/pull/569#issuecomment-1498345361), do you think that can be fixed in the short term?
* I'm not clear on the status of #617. Do you think it's ready for merging now?

**Comment by hyenias** [comment]
For #617 I finished testing it a while ago. The only thing still missing is the small diff I put in the comments at https://github.com/ksh93/ksh/pull/617#issuecomment-1501012591. Then your usual clean-up and standardization stuff. Perhaps you might want to redo src/cmd/ksh93/tests/printf.sh and collapse it down into another existing test file. Lastly, probably an entry needs to be made into the src/cmd/ksh93/COMPATIBILITY log to denote the change in printf's behavior.

**Comment by phidebian** [comment]
Hi All,

I am both remote at the moment (with slow network access trough my phone AccessPoint) and off-sick, so I can not really re-build re-run the tests/ (will be off until June 1st)

Regarding https://github.com/ksh93/ksh/pull/569, I missed your question at the time, sorry about that.

I guess the remaining problem is
````
$ TZ=UTC LC_TIME=C arch/darwin.arm64-64/bin/ksh -c 'printf "%T\n" "now" "24 hours ago" "48 hours ago" "72 hours ago"'
Thu Apr  6 00:52:07 UTC 2023
Thu Apr  6 00:52:07 UTC 2023
Wed Apr  5 00:52:07 UTC 2023
Tue Apr  4 00:52:07 UTC 2023
````
You say `'24 hours ago' is in fact the same as 'now'. All the 'hours ago' seems to be off by 24 hours:`

May be the reason I skipped it is that I don't actually reproduce it, a run on my ubuntu LTS 22.04 gives

````
$ TZ=UTC LC_TIME=C ./arch/gnu.i386-64/bin/ksh -c 'printf "%T\n" "now" "24 hours ago" "48 hours ago" "72 hours ago"'
TZ=UTC LC_TIME=C ./arch/gnu.i386-64/bin/ksh -c 'printf "%T\n" \
"now" \
"24 hours ago" \
"48 hours ago" \
"72 hours ago"'
Tue May 16 03:23:17 UTC 2023
Mon May 15 03:23:17 UTC 2023
Sun May 14 03:23:17 UTC 2023
Sat May 13 03:23:17 UTC 2023
````

So to me looks good.

I pushed back https://github.com/phidebian/ksh/tree/bug-182 the branch, the branch I proposed a while back, and the run above.

For #617, I think it was OK modulo the coding style and various conventions.

Yet @hyenias suggest merging test files, is there an incentives for that? Actually I am in favor of vast multiplication of the test file (=`*.sh`, ideally running only 1 test per .sh, the reason for that is that I use my own test suite, derived from the official one, I got a set of sed like script that massage the `*.sh` so I as much as I can I try to reach the goal of 1 test per `*.sh`, from there my QA run run all in parallel making the original 10 min journey for the QA run to a 10s one, meaning while developing, QA is almost on auto loop I see regression immediately. I also isolated the QA run with long time out, those are run only when the quick and dirty QA run pass, then I run the 10min official QA, generally passing, allowing myself a little break. So if the `*.sh` are splitted on the official source tree, it reduce the burden of maintaining the sed like scripts for splitting them. So unless there are reason I don't see at the moment, why reducing the number of `*.sh` ?
Well in short this is extreme programing :-)

May be one reason could be about the archeologist (I often forgot this dimention) i.e don't upset existing `*.sh` code base so a problem in the QA could be bisected ? but then for the new test that didn't existed in the past may be those could be accepted as multiplied new test files (t-1.sh, t-2.sh, etc...)

Note that I never proposed/submitted my splitted `*.sh` files with the idea it would be too upseting the code base.

Cheers

**Comment by hyenias** [comment]
To prepare for release 1.05, I went testing current ksh head on more of my development boxes. I get lots of various test failures on my FreeBSD boxes both arm64 and amd64 running 13.2 and 13.1-p6.

Would someone be so kind and please test current ksh build on their FreeBSD box(es) and indicate whether all tests pass or not?

**Comment by pghvlaans** [comment]
I test on FreeBSD whenever I do a PR, and they never have for me. I ran the tests on the 1.0 branch just now to check again. The vast majority of failures (possibly all failures) are related to missing `/dev/fd/3` or `/dev/fd/5`; my 13.2 installation only seems to go up to `/dev/fd/2` (created by default, per `fdescfs(5)`).

**Comment by posguy99** [comment]
> To prepare for release 1.05, I went testing current ksh head on more of my development boxes. I get lots of various test failures on my FreeBSD boxes both arm64 and amd64 running 13.2 and 13.1-p6.
> 
> Would someone be so kind and please test current ksh build on their FreeBSD box(es) and indicate whether all tests pass or not?

@hyenias I don't normally run fBSD, but I spun up an 13.2-RELEASE aarch64 VM to test, and I get no build/test failures.  `ksh` does call itself /MOD, which is weird since all I did was clone the repo and build, but nothing other than that.

**Comment by JohnoKing** [comment]
These are the current results I get after running the basic.sh regression tests on FreeBSD 13.1-p6 and Haiku:
```
$ bin/shtests -p basic
#### Regression-testing /usr/home/johno/ksh/arch/freebsd13.i386-64/bin/ksh ####
test basic begins at 2023-05-20+01:02:51
	basic.sh[327]: FAIL: process substitution not working
	basic.sh[334]: FAIL: >() process substitution fails
	basic.sh[345]: FAIL: >() process substitution fails in for loop
	basic.sh[347]: FAIL: process substitution of compound commands not working
cat: /dev/fd/4: No such file or directory
	basic.sh[356]: FAIL: process substitution not working outside for or while loop
	basic.sh[357]: FAIL: process substitution not working in for or while loop
join: /dev/fd/3: No such file or directory
	basic.sh[933]: FAIL: pipeline fails with illegal seek error (expected $'1\n2', got '')
test basic failed at 2023-05-20+01:02:55 with exit code 7 [ 146 tests 7 errors ]
Total errors: 7
```
There are quite a few more test failures than those, which indicates process substitution is completely borked on both of those operating systems. After some debugging, I've determined the bug was introduced in c7363090ece (the changes to the DEVFD probe are most likely responsible for introducing this bug).
Side note: I have determined that the bug contaminates the build directory in such a way that attempting to rebuild on an older commit will produce a broken binary if the build directory wasn't completely cleared out. When recompiling, make sure you completely clear it out with `rm -r ./arch` or `git clean -fdx`.

**Comment by McDutchie** [comment]
> `ksh` does call itself /MOD, which is weird since all I did was clone the repo and build, but nothing other than that.

That happened to me once as well but it hasn't made it to my priority list yet. It should be fixed though. `/MOD` is added when the working directory is not clean, i.e., if it is a git repo with local uncommitted modifications. So apparently the following code to determine whether the working directory is clean is not quite working right: https://github.com/ksh93/ksh/blob/a5cfb049637e2a428480e19529976f1088cd8fd0/src/lib/libast/Mamfile#L47-L50

Does anyone know of a reliable way to determine if the working directory is clean?

**Comment by McDutchie** [comment]
> There are quite a few more test failures than those, which indicates process substitution is completely borked on both of those operating systems.

I cannot reproduce this on FreeBSD. I'm testing ksh on both FreeBSD 12.4-RELEASE-p2 amd64 and FreeBSD 13.1-RELEASE-p7 arm64, and on both, all regression tests pass consistently.

**Comment by pghvlaans** [comment]
Here's something interesting: if built with `sh` rather than `ksh` on FreeBSD, the failures do not occur. That would explain why everything passed on a brand new VM.

**Comment by McDutchie** [comment]
Very interesting. What is the ksh version used?

**Comment by McDutchie** [comment]
> Very interesting. What is the ksh version used?

Well, this is bad. I can reproduce the problem on FreeBSD 13.1 (but not 12.4) if I install a freshly built ksh 93u+m in /usr/local/bin and then rebuild ksh from scratch using SHELL=/usr/local/bin/ksh.

**Comment by pghvlaans** [comment]
I got the fails tonight from `93u+m/1.05-beta 2023-05-19`, but I'd been seeing similar results when something from around 650d415 or 6b0489a on `dev` (don't remember exactly what) was in the path.

**Comment by JohnoKing** [comment]
I am not using ksh to build itself, but rather `sh` on FreeBSD and `bash` on Haiku, yet I still get the bug regardless. I have no clue what is happening there. Additionally, I'm using the latest commit of ksh to build on Linux and don't get the bug there.

~Anyways, I have patched out the process substitution failures on my end by creating the following patch to revert the DEVFD probe change in c7363090, while still keeping the probe in the Mamfile:~ **EDIT:** This patch is incorrect, see https://github.com/ksh93/ksh/issues/619#issuecomment-1555947212
<details>
<summary>Incorrect patch that actually enables the FIFO fallback</summary>

```diff
diff --git a/src/cmd/ksh93/Mamfile b/src/cmd/ksh93/Mamfile
index e0f13928b..9a194ffee 100644
--- a/src/cmd/ksh93/Mamfile
+++ b/src/cmd/ksh93/Mamfile
@@ -60,7 +60,7 @@ make install
 			exec -		'DEVFD=')
 			exec -			# if /dev/fd/n exposes file descriptor n, make SHOPT_DEVFD use this for
 			exec -			# process substitutions instead of FIFOs, which are not as secure
-			exec -			test -e /dev/fd/9 9<&0 && test ! -e /dev/fd/9 9<&- && writedef DEVFD 1 ;;
+			exec -			test -e /dev/fd/9 9<&0; writedef DEVFD $? ;;
 			exec -		'TEST_L=')
 			exec -			# if external 'test' supports lowercase -l as equivalent to -L, enable
 			exec -			# SHOPT_TEST_L so ksh supports it for compatibility with local scripts
```

</details>

**Comment by pghvlaans** [comment]
> > Very interesting. What is the ksh version used?
> 
> Well, this is bad. I can reproduce the problem on FreeBSD 13.1 (but not 12.4) if I install a freshly built ksh 93u+m in /usr/local/bin and then rebuild ksh from scratch using SHELL=/usr/local/bin/ksh.

Everything checks out fine with `oksh`, but `bash` causes the same failures.

**Comment by posguy99** [comment]
> Here's something interesting: if built with `sh` rather than `ksh` on FreeBSD, the failures do not occur. That would explain why everything passed on a brand new VM.

Yea, I am building with the default `sh` on fBSD.

If I build under the `ksh` I just built without any regression test failures, I get the `basic.sh` failures as @JohnoKing does.

**Comment by McDutchie** [comment]
> Anyways, I have patched out the process substitution failures on my end by creating the following patch to revert the DEVFD probe change in [c736309](https://github.com/ksh93/ksh/commit/c7363090ece5c4619cc26be48920328649e18fc2), while still keeping the probe in the Mamfile:

That doesn't revert the change, as the writedef() has an opposite logic to the old option() function in features/options: it takes an argument for #define as an option, not an exit status, so 0 now disables the option. So, since /dev/fd/9 exists with FD 9 open, test returns 0, which causes writedef to disable SHOPT_DEVFD. Which means you're now using the process substitution fallback that uses FIFOs. At least we've confirmed that still works.

Other than that, I still haven't the foggiest clue what's going on. This makes no sense at all.

**Comment by McDutchie** [comment]
Alright, I am starting to make sense of this.

The first problem is not our fault: /dev/fd is b0rken on recent FreeBSD versions.

On FreeBSD 12.4:

```
$ ls -l /dev/fd/9 9</dev/null
cr-xr-xr-x  1 root  wheel  0xc 24 apr. 03:00 /dev/fd/9
```

This is as expected.

But on FreeBSD 13.1:

```
$ ls -l /dev/fd/9 9</dev/null
ls: /dev/fd/9: No such file or directory
```

Second problem: the SHOPT_DEVFD probe in the Mamfile produces a false positive when executed on ksh or bash.

On FreeBSD 13.1:

```
$ /bin/sh -c 'test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rken'
b0rken
$ /usr/local/bin/ksh -c 'echo ${.sh.version}; test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rken'
Version AJM 93u+m/1.0.5-beta 2023-05-19
falsepositive
$ /usr/local/bin/ksh93 -c 'echo ${.sh.version}; test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rken'
Version AJM 93u+ 2012-08-01
falsepositive
$ /usr/local/bin/bash -c 'echo $BASH_VERSION; test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rken'
5.2.15(0)-release
falsepositive
```

/dev/fd/9 doesn't exist due to FreeBSD breakage, yet somehow `test -e` returns a false positive. So, somehow, `test -e` is broken on both ksh and bash! This, of course, breaks the SHOPT_DEVFD probe in the Mamfile.

A workaround might be to use `ls` in the probe instead.

**Comment by posguy99** [comment]
Is this relevant?

```
mwilson@fbsd-virt:~ $ ksh -c 'echo ${.sh.version}; test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rken'
Version AJM 93u+m/1.1.0-alpha+a5cfb049 2023-05-19
falsepositive
```

vs

```
mwilson@fbsd-virt:~ $ ksh -c 'echo ${.sh.version}; /bin/test -e /dev/fd/9 9</dev/null && echo falsepositive || echo b0rk
en'
Version AJM 93u+m/1.1.0-alpha+a5cfb049 2023-05-19
b0rken
```


So the builtin `test` generates the false positive.

**Comment by McDutchie** [comment]
Yes, that is as expected. ksh is special-casing `/dev/fd` here in a way that I think is completely broken (and so is bash). Perhaps they do this because both ksh and bash can use, e.g., `> /dev/fd/9` in redirections even if `/dev/fd/9` doesn't physically exist.

But that doesn't mean you can use it for everything, e.g., external commands. `test -e` should always physically test whether a file exists or not, anything else is just completely broken and violates POSIX.

Unfortunately there may well be scripts that depend on this flawed design to test if they can use `/dev/fd` in redirections, or to test whether a file descriptor is open. A quick google already shows some evidence of that on stackexchange. So a fix for `test` may have to be limited to the POSIX mode. :/

**Comment by McDutchie** [comment]
The following patch fixes the problem on FreeBSD 13.1 by using `ls` for the SHOPT_DEVFD probe. With this, ksh falls back to FIFOs for process substitution and everything works correctly.

<details><summary>Patch to Mamfile</summary>

```diff
diff --git a/src/cmd/ksh93/Mamfile b/src/cmd/ksh93/Mamfile
index e0f13928b..f8ce2ed57 100644
--- a/src/cmd/ksh93/Mamfile
+++ b/src/cmd/ksh93/Mamfile
@@ -60,7 +60,11 @@ make install
 			exec -		'DEVFD=')
 			exec -			# if /dev/fd/n exposes file descriptor n, make SHOPT_DEVFD use this for
 			exec -			# process substitutions instead of FIFOs, which are not as secure
-			exec -			test -e /dev/fd/9 9<&0 && test ! -e /dev/fd/9 9<&- && writedef DEVFD 1 ;;
+			exec -			# * WARNING: do not use 'test -e' for this probe, as some shells special-case
+			exec -			# * 'test -e /dev/fd/9' (in violation of POSIX), turning it into a test for an
+			exec -			# * open file descriptor 9 instead of /dev/fd/9's existence in the file system.
+			exec -			{ ls -d /dev/fd/9 9<&0 && ! ls -d /dev/fd/9 9<&- ; } >/dev/null 2>&1 \
+			exec -				&& writedef DEVFD 1 ;;
 			exec -		'TEST_L=')
 			exec -			# if external 'test' supports lowercase -l as equivalent to -L, enable
 			exec -			# SHOPT_TEST_L so ksh supports it for compatibility with local scripts
```
</details>

The following patch disables the flawed special-casing for /dev/fd/n in `test` and `[[` if POSIX mode is on.

<details><summary>Patch to test.c</summary>

```diff
diff --git a/src/cmd/ksh93/bltins/test.c b/src/cmd/ksh93/bltins/test.c
index 321bc22d8..2bb837d33 100644
--- a/src/cmd/ksh93/bltins/test.c
+++ b/src/cmd/ksh93/bltins/test.c
@@ -607,7 +607,7 @@ int sh_access(const char *name, int mode)
 	struct stat statb;
 	if(*name==0)
 		return -1;
-	if(sh_isdevfd(name))
+	if(!sh_isoption(SH_POSIX) && sh_isdevfd(name))
 		return sh_ioaccess((int)strtol(name+8, NULL, 10),mode);
 	/* can't use access function for execute permission with root */
 	if(mode==X_OK && sh.euserid==0)
```
</details>

**Comment by pghvlaans** [comment]
For what it's worth, this is what the `oksh` code says about `test -e` at `c_test.c:241`:

    	case TO_FILEXST: /* -e */
    		/* at&t ksh does not appear to do the /dev/fd/ thing for
    		 * this (unless the os itself handles it)
    		 */
    		return stat(opnd1, &b1) == 0;

**Comment by McDutchie** [comment]
Yeah, that comment is wrong :)

*edit:* ksh93-history repo confirms it's been doing this since 1995 or before.

**Comment by posguy99** [comment]
> For what it's worth, this is what the `oksh` code says about `test -e` at `c_test.c:241`:

`mksh` has the same code but without the comment.  Logical, they're both based on `pdksh`.

**Comment by hyenias** [comment]
Thank you all for confirming the FreeBSD trouble along with testing in greater detail. @McDutchie I used your two patches and they work for me as well.

For the git modification determination, I always manually run a git status to check where I stand in a directory. So, just use a version of the following:
```
t=$(git status --untracked-files=no --porcelain); e=$?; if (( e==0 )) && [[ -z $t ]]; then echo Not modified; else echo Modified or git error; fi
```

**Comment by pghvlaans** [comment]
> Does anyone know of a reliable way to determine if the working directory is clean?

I'm doing some testing, and it looks like `bin/mamprobe` is triggering `git diff-index` because it gets copied in from another location and the index is not updated before the check for MOD. I'll try to have a PR put together pretty soon.

**Comment by McDutchie** [comment]
According to [this StackExchange answer](https://unix.stackexchange.com/a/394674), a `git update-index --really-refresh` should fix it.

Although we could probably do with removing `bin/mamprobe` from the git repo as well since it will, as you say, get copied in, and it's not needed for direct invocation by users.

```diff
diff --git a/src/lib/libast/Mamfile b/src/lib/libast/Mamfile
index 5704dfd92..b0b359858 100644
--- a/src/lib/libast/Mamfile
+++ b/src/lib/libast/Mamfile
@@ -26,11 +26,12 @@ make install
 					exec -	case $git_commit in
 					exec -	????????)
 					exec -		echo "/* generated by libast/Mamfile using $(git --version) */"
+					exec -		git update-index --really-refresh >/dev/null
 					exec -		case $(git branch 2>/dev/null) in
 					exec -		'' | *\*\ [0-9]*.[0-9]*)
 					exec -			# If we're not on a git branch (tarball) or on a branch that starts
 					exec -			# with a number (release branch), then compile as a release version
-					exec -			if	git diff --quiet	# WD clean?
+					exec -			if	git diff-index --quiet HEAD	# WD clean?
 					exec -			then	echo "#ifndef _AST_release"
 					exec -				echo "#    define _AST_release	1"
 					exec -				echo "#endif"
```

**Comment by pghvlaans** [comment]
Oh, nice. What I've got uses `update-index` with filenames, but your patch doesn't potentially run into problems with different `awk` versions.

**Comment by McDutchie** [comment]
Git fix (+ other tweaks) applied in 1faf04d58 (dev) and 996b2a5f0 (1.0).

**Comment by McDutchie** [comment]
POSIX mode fix for /dev/fd/n in test/[/[[ applied in 293790ab0 (dev) and 3da5338a8 (1.0).

I think the 1.0 branch is just about ready for the 1.0.5 release; no other open bugs are close to a resolution.

One last round of testing... please join in and test the heck out of it :)

**Comment by JohnoKing** [comment]
The build on Haiku is still broken because the Mamfile fix in https://github.com/ksh93/ksh/issues/619#issuecomment-1555957799 hasn't been commited to the dev and 1.0 branches yet. (Haiku uses bash as `/bin/sh`, which has the `/dev/fd` detection problem.)

Besides that, I haven't found any problems yet. I post an update with more in depth testing results later.

**Comment by JohnoKing** [comment]
I've tested the latest 1.0 branch commit and have encountered the following problems:

Problem One: The build on illumos is borked: the regression tests will intermittently lock up in io.sh or fail in places where they don't in 1.0.4. Yet again, the commit at fault is c7363090, thanks to the /dev/fd feature test changes. To be specific, the following command added in c7363090 fails on illumos:
```sh
! ls -d /dev/fd/9 9<&-
```
This causes the FIFO method to be used instead of `/dev/fd`. This wouldn't be much of an issue, if not for the FIFO method causing lockups on illumos for whatever reason + a bizzare set of failures in the sh_match regression tests.
The Mamfile patch in https://github.com/ksh93/ksh/issues/619#issuecomment-1555957799 does *NOT* fix the illumos build because the feature test contained therein still fails.
Running the regression tests on illumos after force-enabling `SHOPT_DEVFD` in SHOPT.sh fixes the lockup in io.sh and sh_match failures.
The `/dev/fd` method works better on illumos than the FIFO method, so I would suggest amending the `SHOPT_DEVFD` feature test to allow illumos to use it.

Problem Two: There is a minor regression in the results for shbench's numfor benchmark (although there is also an improvement in the sample benchmark):
```
(Compiled with CCFLAGS=-O2)
--------------------------------------------------------------------------------
name             /tmp/ksh-1.0.4       /tmp/ksh-1.0.5-pre   /tmp/ksh-1.0.5-revert
--------------------------------------------------------------------------------
braces.ksh       0.183 [0.179-0.196]  0.194 [0.189-0.206]  0.185 [0.178-0.206]
charsplit.ksh    0.415 [0.406-0.435]  0.414 [0.406-0.458]  0.410 [0.405-0.432]
extglob.ksh      0.220 [0.204-0.257]  0.219 [0.206-0.232]  0.221 [0.212-0.247]
fibonacci.ksh    0.209 [0.199-0.249]  0.216 [0.210-0.232]  0.204 [0.198-0.222]
gsub.ksh         0.330 [0.312-0.368]  0.332 [0.314-0.361]  0.333 [0.321-0.354]
hanoi.ksh        0.164 [0.157-0.175]  0.167 [0.161-0.185]  0.158 [0.153-0.167]
numfor.ksh       0.320 [0.309-0.362]  0.412 [0.400-0.458]  0.327 [0.321-0.369]
numloop.ksh      0.245 [0.241-0.291]  0.262 [0.256-0.304]  0.254 [0.251-0.260]
parens.ksh       0.129 [0.125-0.135]  0.132 [0.129-0.137]  0.129 [0.126-0.135]
piln2.ksh        0.549 [0.538-0.577]  0.559 [0.547-0.656]  0.556 [0.540-1.127]
plus-equals.ksh  0.232 [0.225-0.245]  0.239 [0.226-0.325]  0.233 [0.229-0.248]
printf.ksh       0.167 [0.160-0.209]  0.171 [0.163-0.218]  0.170 [0.160-0.232]
qsort.ksh        0.503 [0.488-0.556]  0.521 [0.495-0.594]  0.505 [0.487-0.537]
rand.ksh         0.464 [0.444-0.491]  0.460 [0.445-0.488]  0.463 [0.452-0.491]
rand2.ksh        0.339 [0.327-0.357]  0.332 [0.321-0.359]  0.339 [0.320-0.358]
sample.ksh       0.638 [0.625-0.671]  0.520 [0.503-0.585]  0.509 [0.492-0.558]
shuffle.ksh      0.399 [0.391-0.412]  0.426 [0.405-0.448]  0.396 [0.383-0.417]
sieve.ksh        0.409 [0.395-0.468]  0.456 [0.443-0.477]  0.397 [0.372-0.425]
--------------------------------------------------------------------------------
```
I've identified commit 45b18082 as the cause of the performance regression. As for the results above, ksh-1.0.5-pre is the latest commit on the 1.0 branch, while the ksh-1.0.5-revert is the current 1.0 branch with 45b18082 reverted. I'm doubtful fixing this is feasible though. Older commits are faster because the act of saving and restoring the signal mask is skipped. Reverting 45b18082 carries a significant risk of (re)introducing signal handling bugs.

Problem Three: There are five more regression test failures on DragonFly BSD. I currently lack the free time necessary to investigate this further.

Operating Systems I've tested with no regression test errors on the 1.0 branch:
Arch Linux
Void Linux w/ musl libc
FreeBSD 13.2

Operating systems with regression test failures:
Haiku R1/beta4 hrev56578+86 (58 errors, down from 64 errors in 1.0.4. Without the Mamfile fix in https://github.com/ksh93/ksh/issues/619#issuecomment-1555957799 there are far more test failures, however.)
DragonFly BSD 6.4 (24 errors, up from 19 errors in 1.0.4)
NetBSD 9.3 (5 errors; no change)
OmniOS CE r151046 (Has variation; some failures are intermittent. The FIFO method, which is the default on illumos as of c7363090, causes a lockup and more test failures that the `/dev/fd` method doesn't have.)

Side note: I tested ksh on Cygwin about a week ago and there was no change in the number of regress fails then. I don't have the time to re-test the latest commit right now, though I will do so later when I have the time.

**Comment by McDutchie** [comment]
> Problem Two: There is a minor regression in the results for shbench's numfor benchmark (although there is also an improvement in the sample benchmark):

That's actually a large performance regression in arithmetic for loops (ksh.old is current 1.0 branch, ksh is same with the revert; both compiled on macOS 12.6.3 arm64 with `-Os`):
```
$ time arch/darwin.arm64-64/bin/ksh.old -c 'integer i; for ((i=0;i<10000000; i++)); do :; done'

real	0m02,71s
user	0m02,22s
sys	0m00,48s
$ time arch/darwin.arm64-64/bin/ksh -c 'integer i; for ((i=0;i<10000000; i++)); do :; done'    

real	0m01,33s
user	0m01,22s
sys	0m00,01s
```

It's not limited to arithmetic for loops:

```
$ time arch/darwin.arm64-64/bin/ksh.old -c 'integer i=0; while((++i<10000000)); do :; done'   

real	0m02,44s
user	0m01,96s
sys	0m00,48s
$ time arch/darwin.arm64-64/bin/ksh -c 'integer i=0; while((++i<10000000)); do :; done'    

real	0m00,98s
user	0m00,96s
sys	0m00,01s
```

I am not okay with this. :-/

**Comment by McDutchie** [comment]
Looks like the performance regression is actually not in the loop at all (the `TWH` code in sh_exec() does not save the signal mask as it passes 0 to sigsetjmp). It is running the builtin command in the body, `:` in this case. And after the fix, ksh now saves the signal mask when running a builtin, in the `TCOM` code.

The following patch makes it not do that, and lo and behold, the performance regression is gone:

```diff
diff --git a/src/cmd/ksh93/sh/xec.c b/src/cmd/ksh93/sh/xec.c
index 154ea096b..848bb0aec 100644
--- a/src/cmd/ksh93/sh/xec.c
+++ b/src/cmd/ksh93/sh/xec.c
@@ -1155,7 +1155,7 @@ int sh_exec(const Shnode_t *t, int flags)
 					if(execflg)
 						sh_onstate(SH_NOFORK);
 					sh_pushcontext(buffp,SH_JMPCMD);
-					jmpval = sigsetjmp(buffp->buff,1);
+					jmpval = sigsetjmp(buffp->buff,0);
 					if(jmpval == 0)
 					{
 						if(!(nv_isattr(np,BLT_ENV)))
```

Since ksh seems to have been running builtins just fine without saving the signal mask for decades (the AT&T folks thought they were using sigsetjmp but weren't), I see no evidence that this (re)introduces any bugs.

**Comment by McDutchie** [comment]
@JohnoKing wrote:
> This causes the FIFO method to be used instead of `/dev/fd`. This wouldn't be much of an issue, if not for the FIFO method causing lockups on illumos for whatever reason + a bizzare set of failures in the sh_match regression tests.

I noticed that FIFOs are broken on Solaris and illumos years ago, while testing [modernish](/modernish/modernish). The Solaris folks fixed it after I [complained](https://groups.google.com/g/comp.unix.shell/c/Jq8Wcj_xBPA/m/1_tUvVRPAgAJ) about it on Usenet. So I submitted a [bug report](https://www.illumos.org/issues/10179) to illumos on 2019-01-05 in which I even included an explanation I obtained from Casper Dik at Oracle detailing how they had fixed it in Solaris. And there it sits. The illumos folks have ignored it completely.

**Comment by McDutchie** [comment]
In any case, the build should now be fixed on both illumos and FreeBSD 13.1 (as of 504d645f9e7bed9bdccb295c99b620f1f3bf3655 (dev) and b5f433d375b979277c6fe664e61fef583367b8ea (1.0)).

**Comment by JohnoKing** [comment]
I have been able to test the latest ksh commit again and have obtained the following results:
All the regression tests now pass on FreeBSD without issue.
On illumos, ksh is working properly again. [These tests in signal.sh](https://github.com/ksh93/ksh/blob/504d645f9e7bed9bdccb295c99b620f1f3bf3655/src/cmd/ksh93/tests/signal.sh#L268-L307) fail intermittently, although that is a pre-existing issue which also exists in 1.0.4.
On Cygwin, there are currently 207 test errors (down from 208 in 1.0.4). The majority of these (such as the ones in arith.sh) are probably caused by Cygwin itself and are pre-existing issues.
On DragonFly BSD, the new regression test failures are symptoms of #432 (cf. efa18c85).

There are no regressions here that would prevent a release; everything that remains are pre-existing bugs not introduced in 1.0.5.

**Comment by pghvlaans** [comment]
I'm wrapping up testing for the SlackBuild patch I'll be requesting on LinuxQuestions (just the ARM port on a docker image left, which is always a bit time-consuming). Highlights:

* Slackware-standard O2 optimization for a performance boost (about 10-15% when I last checked a few months ago)
* Copyright file
* Install the functions from fun and include a profile.d script for FPATH

I don't know if the build will be patched this time around, but I'm hopeful the new version will be added to -current when it's ready in any case.

**Comment by McDutchie** [comment]
Done in 9b251344b45cd6250a6588435c0717aee8dd8298.
